### ISSUE TITLE
On hierarchical connections vs. port mappings (and related problems with properties, bindings and etc.)

### ISSUE THREAD

**Issue by buzden** [saeaadl/aadlv3#15]
This is an issue to try avoid some problems of AADLv2 in AADLv3. Also, it touches some discussion in the committee regarding to very old Peter proposal of considering hierarchical connections to be not as connections as sibling ones.

Let's for instance consider `Latency` property which is applicable to connections. Also, let's consider a two systems `prod` and `cons` each with subsystems `innerProd` and `innerCons`:
```
system Producer
features
  p: out event data port;
end Producer;

system Consumer -- inverse of Producer ;-)
features
  p: in event data port;
end Consumer;

system implementation Producer.outer
subcomponents
  innerProd: system Producer;
connections
  hierarchicalProd: port innerProd.p -> p;
end Producer.outer

system implementation Consumer.outer
subcomponents
  innerCons: system Consumer;
connections
  hierarchicalCons: port p -> p.innerCons;
end Consumer.outer

system container end container;
system implementation container.impl
subcomponents
  prod: system Producer.outer;
  cons: system Consumer.outer;
connections
  siblingProdCons: port prod.p -> cons.p;
end container.impl;
```
This model contains three connections:
- two hierarchical: `hierarchicalProd` and `hierarchicalCons`;
- one sibling: `siblingProdCons`.

According to the part 9.2 of the standard, all these three form a single _semantic connection_.

Let's consider we want to assign the `Latency` property to this semantic connection. Which element should we assign this property to? Well, let's choose the sibling one since sibling connections are always alone in semantic connections. Looks good? Maybe. It really looks good because hierarchical connections in this example does not serve anything but mapping interface of each outer component to its implementation.

But how should we understand the model if both `siblingProdCons` and e.g. `hierarchicalProd` connections have the `Latency` property association? Should their latencies be added to form the latency of the whole semantic connection? Or should they always be the same because they relate to the same semantic connection? So, what is the _meaning_ of `Latency` property association for a hierarchical connection? I can say there are at least two variants of understanding I met:
- property of each connection is a _contribution_ of this particular connection to the latency of the whole semantic connection;
- property of each (hierarchical) connection is an _expectation_ of a component containing this hierarchical connection to the latency of the whole semantic connection.

The second meaning more or less conforms to the understanding of hierarchical connections to be just mappings of implementation to interface. But what if hierarchical connections sometimes are not just a mappings (as they definitely are sometimes) but imply some internal work (e.g. of an operating system or devices for buffering, time alignment, checks and etc.). In this case hierarchical connections
- are not just mappings;
- definitely can contribute to the latency of the whole semantic connection.

---

The same analogously can be said to, e.g. bindings: whether bindings of the sibling connection defines bindings of the whole semantic connection or not? What if hierarchical connection (when being not just a mapping) is provided by some components (which can be expressed with bindings), which are other than a sibling connection? Will it mean that bindings of the whole semantic connection is a union of bindings of separate connections (as sets)? What if there is an intersection in these sets, should then bindings be understood as multisets or not?

---

So, to sum up:
- there is definitely a need to distinguish _hierarchical connections_ (or whatever you call them; they actually do something and can be assigned with all properties that are applicable to connections) and _port mappings_ (which are not connections in terms of action of movement of data or control);
- we really should reconsider notion of semantic connection; probably, we should make them more explicit, probably we can reuse flows notion to cover everything we need from semantic connections.

**Comment by buzden** [comment]
One more thought regarding this topic and also semantic connections, it is the following.

**Semantic connection** is a *sequence of adjacent connections* from an ultimate source to an ultimate destination. The question is *how many* and *what kind* of connections should be in a sequence to treat iе like a semantic connection? (By the way, we are considering not only *fully defined* models, but also *partially defined* too, e.g. with no thread components inside processes and etc.)

First of all, we think that generally semantic connections can be contained of only a single connection, so this is the answer to "how many".

Regarding to "what kind", our first feeling was that if such a sequence of adjacent connections contains a *sibling connection* that the sequence can be treated as a semantic connection. This can be reformulated as "these must exist at least one connection that is *not hierarchical* to be able to treat a sequence as a semantic connection".

But if to distinguish *hierarchical connections* and *port mappings* as this issue supposes, we can consider another meaningful class of semantic connections, not covered by the rule above. That is, the rule can be reformulated to "in semantic connection at least one connection must be not *port mapping*". And this looks like somewhat nice and sound.

**Comment by reteprelief** [comment]
Denis,

I always liked the distinction between a (sibling) connection and a feature mapping (up or down connection).  Simplifies the definition of a semantic connection between components. 

We can also define an "external connection", i.e., a connection from a component within the system to the external world - it consists of a sequence of feature mappings from/to a feature of the top-level component. They come in handy when dealing with fault propagation for example.
